runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,424 @@
+
+        Collection<CorfuRecord<Val1, Meta1>> filterResult1 = scanAndFilter(tableName1, queryOptions1.getTimestamp(), query1);
+        Collection<R> queryResult1 = transform(
+                filterResult1,
+                initializeResultCollection(queryOptions1),
+                queryOptions1.getProjection());
+
+        Collection<CorfuRecord<Val2, Meta2>> filterResult2 = scanAndFilter(tableName2, queryOptions2.getTimestamp(), query2);
+        Collection<S> queryResult2 = transform(
+                filterResult2,
+                initializeResultCollection(queryOptions2),
+                queryOptions2.getProjection());
+
+        Collection<T> joinResult = new ArrayList<>();
+
+        for (R value1 : queryResult1) {
+            for (S value2 : queryResult2) {
+                if (!joinPredicate.test(value1, value2)) {
+                    continue;
+                }
+
+                T resultValue = joinFunction.apply(value1, value2);
+                joinResult.add(resultValue);
+            }
+        }
+
+        return new QueryResult<>(joinResult.stream()
+                .map(v -> Optional.ofNullable(joinProjection)
+                        .map(function -> function.apply(v))
+                        .orElse((U) v))
+                .collect(Collectors.toList()));
+    }
+
+    /**
+     * Merge Function which combines the result two tables at a time.
+     *
+     * @param <R> Type of result.
+     */
+    @FunctionalInterface
+    public interface MergeFunction<R> {
+
+        /**
+         * Merges the specified argument list.
+         *
+         * @param arguments Objects across tables to be merged.
+         * @return Resultant object.
+         */
+        R merge(List<Object> arguments);
+    }
+
+    /**
+     * Combines the result of a multi table join, two tables at a time
+     * @param list
+     * @param mergePayload
+     * @param func
+     * @param depth
+     * @param <R>
+     * @return
+     */
+    private <R> List<R> merge(@Nonnull List<Collection<?>> list,
+                              @Nonnull List<Object> mergePayload,
+                              @Nonnull MergeFunction<R> func,
+                              int depth) {
+        Collection<?> collection = list.get(depth);
+        if (list.size() - 1 == depth) {
+            List<R> result = new ArrayList<>();
+            for (Object o : collection) {
+                List<Object> finalMergeList = new ArrayList<>(mergePayload);
+                finalMergeList.add(o);
+                R mergeResult = func.merge(finalMergeList);
+                if (mergeResult != null) {
+                    result.add(mergeResult);
+                }
+            }
+            return result;
+        }
+
+        List<R> result = new ArrayList<>();
+        for (Object o : collection) {
+            List<Object> mergeList = new ArrayList<>(mergePayload);
+            mergeList.add(o);
+            result.addAll(merge(list, mergeList, func, depth + 1));
```

Comment:
not sure if a recursive call a good idea

Reply:
again this is just a first cut. on top of this commit, there will be performance tests and correctness tests where we will likely revisit these changes.